lib/src/default_revset_engine.rs

```diff
@@ -491,25 +492,33 @@ impl<'index, I1: Iterator<Item = IndexEntry<'index>>, I2: Iterator<Item = IndexE
     }
 }
 
+// TODO: Having to pass both `&dyn Index` and `CompositeIndex` is a bit ugly.
+// Maybe we should make `CompositeIndex` implement `Index`?
```

Comment:
random thought: if `CompositeIndex`, `&dyn Index`, etc. were somehow abstracted over owned `Arc<ReadonlyIndex>` and `&MutableIndex`, an evaluated revset and change_id_index might be able to be free from the `'index` lifetime (for `ReadonlyRepo`.) It will remove some unsafe code.

(no action needed)

Reply:
That sounds like a good idea. I'll try to find time to experiment with that later. Thanks.